File: app/src/main/java/com/ed/redditapp/lib/AsyncTask.java
package com.ed.redditapp.lib;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class AsyncTask {
    public static class UIHandler {
        private final Handler handler;

        public UIHandler() {
            this.handler = new Handler(Looper.getMainLooper());
        }

        public void onUIThread(Runnable r) {
            handler.post(r);
        }
    }

    public interface Task {
        void run(UIHandler h);
    }

    public void runAsync(Task task) {
        UIHandler handler = new UIHandler();
        ExecutorService executor = Executors.newSingleThreadExecutor();
        executor.execute(() -> {
            task.run(handler);
        });
    }
}

File: app/src/main/java/com/ed/redditapp/lib/AsyncTask.kt
package com.ed.redditapp.lib

import android.os.Handler
import android.os.Looper
import java.util.concurrent.Executors

typealias AsyncTaskRunnable = (h: AsyncTask.UIHandler) -> Unit

class AsyncTask {
    class UIHandler {
        private val handler = Handler(Looper.getMainLooper())
        fun onUIThread(r: Runnable) = handler.post(r)
    }

    fun runAsync(task: AsyncTaskRunnable) {
        val handler = UIHandler()
        val executor = Executors.newSingleThreadExecutor()
        executor.execute { task(handler) }
    }
}